packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -38,13 +47,44 @@ class UpgradeCommand extends FlutterCommand {
     }
 
     final FlutterVersion flutterVersion = FlutterVersion.instance;
+    final GitTagVersion gitTagVersion = GitTagVersion.determine();
+    if (!argResults['force'] && gitTagVersion == const GitTagVersion.unknown()) {
+      // If the commit is a recognized branch and not master,
+      // explain that we are avoiding potential damage.
+      if (flutterVersion.channel != 'master' && FlutterVersion.officialChannels.contains(flutterVersion.channel)) {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. It is recommended to use git directly if not working off of '
+          'an official channel.'
+        );
+      // Otherwise explain that local changes can be lost.
+      } else {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. If it is okay to remove local changes, then re-run this '
+          'command with --force.'
+        );
+      }
+    }
+    bool stashedChanges = true;
+    final String stashName = 'flutter-upgrade-from-v${gitTagVersion.x}.${gitTagVersion.y}.${gitTagVersion.z}';
+    try {
+      final RunResult runResult = await runCheckedAsync(<String>[
+        'git', 'stash', 'push', '-m', stashName
+      ]);
+      if (runResult.stdout.contains('No local changes to save')) {
```

Comment:
Are we sure that all Git version use this same string

Reply:
No, but we can check for the message in the return string.